src/Microsoft.Extensions.PlatformAbstractions/Native/PlatformApis.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using System.IO;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.Extensions.PlatformAbstractions.Native
+{
+    internal static class PlatformApis
+    {
+        private class DistroInfo
+        {
+            public string Id;
+            public string VersionId;
+        }
+
+        private static readonly Lazy<Platform> _platform = new Lazy<Platform>(DetermineOSPlatform);
+        private static readonly Lazy<DistroInfo> _distroInfo = new Lazy<DistroInfo>(LoadDistroInfo);
+
+        public static string GetOSName()
+        {
+            switch (GetOSPlatform())
+            {
+                case Platform.Windows:
+                    return "Windows";
+                case Platform.Linux:
+                    return GetDistroId();
+                case Platform.Darwin:
+                    return "Mac OS X";
+                default:
+                    return "Unknown";
+            }
+        }
+
+        public static string GetOSVersion()
+        {
+            switch (GetOSPlatform())
+            {
+                case Platform.Windows:
+                    return NativeMethods.Windows.RtlGetVersion();
+                case Platform.Linux:
+                    return GetDistroVersionId();
+                case Platform.Darwin:
+                    return GetDarwinVersion();
+                default:
+                    return string.Empty;
+            }
+        }
+
+        private static string GetDarwinVersion()
+        {
+            var kernelRelease = NativeMethods.Darwin.GetKernelRelease();
+            var version = Version.Parse(kernelRelease);
+            if (version.Major < 5)
+            {
+                // 10.0 covers all versions prior to Darwin 5
+                return "10.0";
+            }
+            else
+            {
+                // Mac OS X 10.1 mapped to Darwin 5.x, and the mapping continues that way
+                // So just subtract for from the Darwin version.
+                // https://en.wikipedia.org/wiki/Darwin_%28operating_system%29
+                return $"10.{version.Major - 4}";
+            }
+        }
+
+        public static Platform GetOSPlatform()
+        {
+            return _platform.Value;
+        }
+
+        private static string GetDistroId()
+        {
+            return _distroInfo.Value?.Id;
+        }
+
+        private static string GetDistroVersionId()
+        {
+            return _distroInfo.Value?.VersionId;
+        }
+
+        private static DistroInfo LoadDistroInfo()
+        {
+            try
+            {
+                var lines = File.ReadAllLines("/etc/os-release");
```

Comment:
You're going to return null regardless if the file doesn't exist. Do we know that everywhere we run that the file exists? I don't see why we can't do this simple check to avoid an exception we know about.